src/GZCTF/Repositories/GameRepository.cs

```diff
@@ -142,211 +143,219 @@ public Task<Game[]> GetGames(int count, int skip, CancellationToken token) =>
-                                        SubmissionType.FirstBlood => Convert.ToInt32(
-                                            s.GameInstance.Challenge.CurrentScore *
-                                            game.BloodBonus.FirstBloodFactor),
-                                        SubmissionType.SecondBlood => Convert.ToInt32(
-                                            s.GameInstance.Challenge.CurrentScore *
-                                            game.BloodBonus.SecondBloodFactor),
-                                        SubmissionType.ThirdBlood => Convert.ToInt32(
-                                            s.GameInstance.Challenge.CurrentScore *
-                                            game.BloodBonus.ThirdBloodFactor),
-                                        SubmissionType.Normal => s.GameInstance.Challenge.CurrentScore,
-                                        _ => throw new ArgumentException(nameof(status))
-                                    }
-                            };
-                        }).ToList()
-                };
-            }).OrderByDescending(j => j.Score).ThenBy(j => j.LastSubmissionTime) //成绩倒序，最后提交时间正序
-            .Select((j, i) =>
-            {
-                j.Rank = i + 1;
+        float[] bloodFactors =
+        [
+            game.BloodBonus.FirstBloodFactor,
+            game.BloodBonus.SecondBloodFactor,
+            game.BloodBonus.ThirdBloodFactor
+        ];
 
-                if (j.Organization is null)
-                    return j;
+        foreach (var item in submissions)
+        {
+            var challenge = challenges[item.Id];
+            var scoreboardItem = items[item.ParticipantId];
 
-                if (ranks.TryGetValue(j.Organization, out var rank))
+            // 4.1. generate bloods
+            if (challenge.Bloods.Count < 3)
+            {
+                item.Type = challenge.Bloods.Count switch
                 {
-                    j.OrganizationRank = rank + 1;
-                    ranks[j.Organization]++;
-                }
-                else
+                    0 => SubmissionType.FirstBlood,
+                    1 => SubmissionType.SecondBlood,
+                    2 => SubmissionType.ThirdBlood,
+                    _ => SubmissionType.Normal // should not happen
+                };
+                challenge.Bloods.Add(new Blood
+                {
+                    Id = scoreboardItem.Id,
+                    Avatar = scoreboardItem.Avatar,
+                    Name = scoreboardItem.Name,
+                    SubmitTimeUtc = item.SubmitTimeUtc
+                });
+            }
+
+            // 4.2. update score
+            item.Score = noBonus
+                ? item.Type switch
                 {
-                    j.OrganizationRank = 1;
-                    ranks[j.Organization] = 1;
+                    SubmissionType.Unaccepted => 0, // should not happen
```

Comment:
```suggestion
                    SubmissionType.Unaccepted => throw new UnreachableException(),
```

Reply:
I'd prefer the previous one, when something unexpected happened has been saved to db will not cause the generation failure of the scoreboard

Reply:
here we run the query with Where to ensure this, this should be fine?